rfd/0056-sql-backend.md

```diff
@@ -0,0 +1,110 @@
+---
+authors: Jim Bishopp (jim@goteleport.com)
+state: draft
+---
+
+# RFD 56 - SQL Backend
+
+
+## What
+
+A SQL Backend extends Teleport with the ability to store its data in either
+PostgreSQL or CockroachDB with support for self-hosted and cloud-hosted 
+configurations.
+
+
+## Why
+
+Supporting a SQL Backend reduces onboarding time for Teleport customers by
+allowing them to use existing infrastructure.
+
+
+## Scope
+
+This RFD focuses on implementing a SQL Backend for PostgreSQL and CockroachDB
+that is either self-hosted or cloud-hosted. The implementation's design will
+be extensible to allow future work that supports other SQL databases such as
+MySQL.
+
+Cloud-hosted configurations will support AWS RDS/Aurora/Redshift, and GCP Cloud SQL.
+
+The implementation will support connecting to a single endpoint. Failover to
+a secondary endpoint will not be supported but may be considered in a future
+proposal.
+
+
+## Authentication
+
+Self-hosted configurations will require mutual TLS authentication using
+user-generated client certificates. The user's CA, client certificate, and
+client key must be added to the Teleport configuration file. Users must ensure
+the provided CA is trusted by the host where the Teleport authentication server
+is running.
+
+AWS and GCP cloud-hosted configurations require IAM. Teleport uses the
+default credential provider for both AWS and GCP to authenticate using IAM.
+
+Mutual TLS authentication is optionally supported for GCP Cloud SQL.
+Paths to a client certificate and key must be added to the Teleport
+configuration file to enable mTLS.
+
+
+## UX
+
+Teleport users must first configure the instance and database where Teleport will 
+store its data. A new database instance and user must be created. The new user
```

Comment:
Can Teleport create the database itself (given enough permissions)? E.g. for DynamoDB it creates the table itself.

Reply:
@r0mant I was concerned users may already have a database with the same name, for whatever reason, and wanted to provide some flexibility. What about making it optional with a default value of "teleport"?

Reply:
I added the default. LMK if you want me to remove it altogether.